packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,301 @@
+					const absFilePath = path.join(assetDirectory, manifestEntry[0]);
+					return {
+						name: manifestEntry[0],
+						hash: manifestEntry[1].hash,
+						contents: (await readFile(absFilePath)).toString("base64"),
+						metadata: {
+							contentType: getType(absFilePath) || "application/octet-stream",
+						},
+					};
+				})
+			);
+
+			try {
+				const res = await fetchResult<UploadResponse>(
+					`/accounts/${accountId}/workers/assets/upload`,
+					{
+						method: "POST",
+						headers: {
+							"Content-Type": "application/x-ndjson",
+							Authorization: `Bearer ${initializeAssetsResponse.jwt}`,
+						},
+						body: payload.map((x) => JSON.stringify(x)).join("\n"),
+					}
+				);
+				logger.info(
+					`Uploaded bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}`
+				);
+				return res;
+			} catch (e) {
+				if (attempts < MAX_UPLOAD_ATTEMPTS) {
+					logger.info(
+						chalk.dim(
+							`Bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length} upload failed. Retrying...\n`,
+							e
+						)
+					);
+					// Exponential backoff, 1 second first time, then 2 second, then 4 second etc.
+					await new Promise((resolvePromise) =>
+						setTimeout(resolvePromise, Math.pow(2, attempts) * 1000)
+					);
+					if (e instanceof APIError && e.isGatewayError()) {
+						// Gateway problem, wait for some additional time and set concurrency to 1
+						queue.concurrency = 1;
+						await new Promise((resolvePromise) =>
+							setTimeout(resolvePromise, Math.pow(2, gatewayErrors) * 5000)
+						);
+						gatewayErrors++;
+						// only count as a failed attempt after a few initial gateway errors
+						if (gatewayErrors >= MAX_UPLOAD_GATEWAY_ERRORS) {
+							attempts++;
+						}
+					} else {
+						attempts++;
+					}
+					return doUpload();
+				} else if (isJwtExpired(initializeAssetsResponse.jwt)) {
+					throw new FatalError(
+						`Asset upload took too long on bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}. Please try again.`
+					);
+				} else {
+					throw e;
+				}
+			}
+		};
+		// add to queue and run it if we haven't reached concurrency limit
+		void queue.add(() =>
+			doUpload().then((res) => {
+				bucketUploadCount++;
+				completionJwt = res.jwt || completionJwt;
+			})
+		);
+	}
+	queue.on("error", (error) => {
+		logger.error(error.message);
+		throw error;
+	});
+	await queue.onIdle();
+
+	// if queue finishes without receiving JWT from asset upload service (AUS)
+	// AUS only returns this in the final bucket upload response
+	if (!completionJwt) {
+		throw new FatalError(
+			"Failed to receive completion signal for file upload. Please try again.",
+			1
+		);
+	} else if (bucketUploadCount !== initializeAssetsResponse.buckets.length) {
+		throw new FatalError(
+			"Upload completion signal received unexpectedly early - we cannot confirm if all files were successfully uploaded. Please try again.",
+			1
+		);
+	}
```

Comment:
I think because you iterate over the full set of buckets and just update the value of completionJwt as you go rather than early exiting, this will never actually fire. That's fine :) Just flagging in case you didn't realize.
```suggestion
	} else if (bucketUploadCount !== initializeAssetsResponse.buckets.length) {
		throw new FatalError(
			"Asset uploading finished unexpectedly early — we cannot confirm if all files were successfully uploaded. Please try again.",
			1
		);
	}
```

Reply:
whoops that was silly, thanks